[Pipelining] Add to/from CSV format and improved __repr__

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129810
* #129786
* __->__ #129264

_Action.__repr__ gets rearranged so it doesn't require an underscore or
a 's' prefix, but still keeps multi-digit stage and microbatch indices
separated by an alpha character indicating the action type.

to/from CSV methods allow dumping a generated schedule to CSV format for
offline visualization or manual editing in a spreadsheet and reloading
to use at runtime.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225 @chauhang @d4l3k

wconstab committed:
Update

[ghstack-poisoned]

H-Huang commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -612,6 +648,78 @@ def __init__(
+                writer.writerow(self.pipeline_order[rank])
+
+    def _validate_schedule(self):
+        def _validate_rank_actions(
```
This is pretty good to validate for a single rank, but https://github.com/pytorch/pytorch/blob/main/test/distributed/pipelining/test_schedule.py#L453-L554 tries to do it across all ranks (its pretty ugly though) and also doesn't include W.

So we can probably try to consolidate these together later

wconstab replied:
ah, bummer i forgot about that.  lets consolidate.

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

H-Huang reviewed (approved):
Looks good

H-Huang commented:
@pytorchbot merge

wconstab commented:
@pytorchbot merge

wconstab committed:
Update

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchhbot merge

wconstab commented:
@pytorchbot merge

H-Huang commented:
@pytorchbot merge

kwen2501 commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -40,14 +42,49 @@ def __str__(self):
-        return f"{self.computation_type}{self.microbatch_index}_s{self.stage_index}"
+        return f"{self.stage_index}{self.computation_type}{self.microbatch_index}"
+
+    @staticmethod
+    def from_str(str):
+        """Reverse of __repr__"""
+        if match := _action_regex.match(str):
+            stage_index, computation_type, microbatch_index = match.groups()
+            return _Action(
+                _ComputationType.from_str(computation_type),
+                int(microbatch_index),
+                int(stage_index),
+            )
+        elif str == "":
+            return None
+        raise RuntimeError(
+            f"Invalid action string: {str}, should be formatted as [stage][action type][microbatch] e.g. 2F0"
+        )
```
It seems this method defines a contract about the input format. Can the contract be put upfront in the method's description? Words in the `RuntimeError` would work.

H-Huang replied:
Updated the method description

kwen2501 commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -612,6 +649,88 @@ def __init__(
+            for rank in self.pipeline_order:
+                writer.writerow(self.pipeline_order[rank])
+
+    def _validate_schedule(self):
+        # TODO(whc) this should be merged with the logic in test_schedule.py#L453-L554
```
nit: the `_validate_schedule` method could seem a bit heavy to be in a base class -- is there a chance to move it to a separate utility and be in functional form, e.g. `validate(schedule)`?

Another consideration is whether every schedule should support this method -- are we now rewriting every schedule to have a field like `pipeline_order`?

wconstab replied:
I intended to put this in the schedule multi class, since currently that is the one that implement step via pipeline order. If I put it in the base class instead we can move it.

I actually did think about eventually moving more schedules to this format but for now it's evolving.

H-Huang committed:
Update on "[Pipelining] Add to/from CSV format and improved __repr__"


_Action.__repr__ gets rearranged so it doesn't require an underscore or
a 's' prefix, but still keeps multi-digit stage and microbatch indices
separated by an alpha character indicating the action type.

to/from CSV methods allow dumping a generated schedule to CSV format for
offline visualization or manual editing in a spreadsheet and reloading
to use at runtime.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225 chauhang d4l3k

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge

H-Huang commented:
@pytorchbot merge -i
